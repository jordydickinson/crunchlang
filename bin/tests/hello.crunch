extern("C")
fun puts!(str: uint8*): int32 = "puts";

fun main!() {
    var msg: uint8[] = {
        // "Hello "
        72, 101, 108, 108, 111, 32,
        // "world\n\0"
        119, 111, 114, 108, 100, 10, 0
    };

    var _ = puts!(msg as uint8*);
}
