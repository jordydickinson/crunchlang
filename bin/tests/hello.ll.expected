; ModuleID = 'hello'
source_filename = "hello"

@llvm.global_ctors = appending global [1 x void ()*] [void ()* @init.ctors]

declare i32 @puts(i8*)

define void @main() {
entry:
  br label %body

body:                                             ; preds = %entry
  %msg = alloca [13 x i8]
  store [13 x i8] c"Hello world\0A\00", [13 x i8]* %msg
  %_ = alloca i32
  %msg.0 = getelementptr [13 x i8], [13 x i8]* %msg, i32 0, i32 0
  %calltmp = call i32 @puts(i8* %msg.0)
  store i32 %calltmp, i32* %_
  ret void
}

define void @init.ctors() {
entry:
  ret void
}
